// powersync/dist/commands/instance/config.js
import { Command, ux } from '@oclif/core';
import * as PowerSyncClient from '../../service/powersync.client.js';
import { checkThatEnvVariablesAreSet, red } from '../../utils.js';
export default class InstanceConfig extends Command {
    static description = 'Display the instance config.';
    static examples = ['<%= config.bin %> <%= command.id %>'];
    async run() {
        checkThatEnvVariablesAreSet(this.config.bin, true);
        const config = await PowerSyncClient.getInstanceConfig();
        if (!config) {
            return this.log(red('No instance config found.'));
        }
        const formattedConfig = {
            auth: config.config?.client_auth,
            // Assume there is only a single connection
            connection: config.config?.replication?.connections[0],
            id: config.id,
            name: config.name,
            region: config.config?.region
        };
        ux.styledObject(formattedConfig, ['id', 'name', 'region', 'connection', 'auth']);
    }
}
