// powersync/dist/commands/instance/set.js
import { Command, Flags, ux } from '@oclif/core';
import inquirer from 'inquirer';
import * as PowerSyncClient from '../../service/powersync.client.js';
import { ENV_VARIABLES } from '../../service/powersync.client.js';
import { bold, checkThatEnvVariablesAreSet, green, updateEnvFileVariableValue } from '../../utils.js';
export default class InstanceSet extends Command {
    static description = 'Set the current instance.';
    static examples = ['<%= config.bin %> <%= command.id %>', '<%= config.bin %> <%= command.id %> --instanceId=1234'];
    static flags = {
        instanceId: Flags.string({ char: 'i', description: 'Instance ID to change to.' })
    };
    async run() {
        checkThatEnvVariablesAreSet(this.config.bin);
        const { flags } = await this.parse(InstanceSet);
        const list = await PowerSyncClient.getInstancesList();
        const instances = list.instances.map((instance) => ({ name: instance.name, value: instance.id }));
        const questions = [
            {
                choices: instances,
                message: 'Which instance would you like to change to?',
                name: 'instanceId',
                type: 'list',
                when: !flags.instanceId
            }
        ];
        const answers = await inquirer.prompt(questions);
        const instanceId = flags.instanceId || answers.instanceId;
        updateEnvFileVariableValue(ENV_VARIABLES.INSTANCE_ID, instanceId);
        ux.log(green(`Current instance has been set to ${bold(instanceId)}`));
    }
}
